File: 2188-minimized-maximum-of-products-distributed-to-any-store/minimized-maximum-of-products-distributed-to-any-store.cpp
class Solution{
private:
bool check(vector<int>& quantities , int n , int mid)
{
    int count = 0;
    for(auto a : quantities)
    {
        if(a%mid == 0)count += (a/mid);
        else count += (a/mid)+1;
        if( count > n) return false;
    }
    return true;
}

public:
    int minimizedMaximum(int n, vector<int>& quantities) {
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        cout.tie(NULL);
        int lo = 1;
        int hi = *max_element(quantities.begin() , quantities.end());
        int mid;
        int ans;
       
        while(lo <= hi)
        {
            mid = lo + (hi-lo)/2;
            if( check(quantities , n , mid))
            {
                ans = mid;
                hi = mid-1;
            }
            else
            {
                lo = mid+1;
            }
        }
        return ans; 
    }
};

File: 2188-minimized-maximum-of-products-distributed-to-any-store/minimized-maximum-of-products-distributed-to-any-store.java
class Solution {
    
    public boolean isValid(int[] quantities,int stores,double productCap)
    {
        int storesServed=0;
        for(int i=0;i<quantities.length;i++)
        {
            storesServed+=Math.ceil(quantities[i]/productCap);
            if(storesServed>stores)
            return false;
        }
        return true;
    }

    public int minimizedMaximum(int n, int[] quantities) 
    {
        int end = Arrays.stream(quantities).max().orElse(0);
        int start=1;
        int result=-1;
        while(start<=end)
        {
            int mid=start+(end-start)/2;
            if(isValid(quantities,n,mid))
            {
                result=mid;
                end=mid-1;
            }
            else
                start=mid+1;
        }  
        return result;
    }
}